homeassistant/components/caldav/todo.py

```diff
@@ -92,3 +111,71 @@ async def async_update(self) -> None:
             for resource in results
             if (todo_item := _todo_item(resource)) is not None
         ]
+
+    async def async_create_todo_item(self, item: TodoItem) -> None:
+        """Add an item to the To-do list."""
+        try:
+            await self.hass.async_add_executor_job(
+                partial(
+                    self._calendar.save_todo,
+                    summary=item.summary,
+                    status=TODO_STATUS_MAP_INV.get(
+                        item.status or TodoItemStatus.NEEDS_ACTION, "NEEDS-ACTION"
+                    ),
+                ),
+            )
+        except (requests.ConnectionError, DAVError) as err:
+            raise HomeAssistantError(f"CalDAV save error: {err}") from err
+
+        self.async_schedule_update_ha_state(force_refresh=True)
+
+    async def async_update_todo_item(self, item: TodoItem) -> None:
+        """Update a To-do item."""
+        uid: str = cast(str, item.uid)
+        try:
+            todo = await self.hass.async_add_executor_job(
+                self._calendar.todo_by_uid, uid
+            )
+        except NotFoundError as err:
+            raise HomeAssistantError(f"Could not find To-do item {uid}: {err}") from err
+        except (requests.ConnectionError, DAVError) as err:
+            raise HomeAssistantError(f"CalDAV lookup error: {err}") from err
+        vtodo = todo.icalendar_component  # type: ignore[attr-defined]
+        if item.summary:
+            vtodo["summary"] = item.summary
+        if item.status:
+            vtodo["status"] = TODO_STATUS_MAP_INV.get(item.status, "NEEDS-ACTION")
+        try:
+            await self.hass.async_add_executor_job(
+                partial(
+                    todo.save,
+                    no_create=True,
+                    obj_type="todo",
+                ),
+            )
+        except (requests.ConnectionError, DAVError) as err:
+            raise HomeAssistantError(f"CalDAV save error: {err}") from err
+        self.async_schedule_update_ha_state(force_refresh=True)
```

Comment:
I wonder if these should be scheduled or not. Generally we make sure that when a service call finishes, the state is updated. That can be done automatically by HA when `should_poll = True`, but that's not the case with todo I think?

Reply:
Happy to follow a best practice here -- I wasn't aware there was service call behavior to do that. Generally, all todo integrations do need to update state after a mutation, and i was handling it manually, but without blocking the result. Have a pointer to where this happens so i can dig a little more?

Reply:
It's handled by the call service helper. Highlighted code is the case for a single entity. Similar code exists below it for multiple.

https://github.com/home-assistant/core/blob/dev/homeassistant/helpers/service.py#L877-L886

Reply:
Thanks, i have added tests to verify the behavior here. I will audit some of the other integrations and make sure I handled this right there too. I think I was missing the "should_poll" part as being the magic which confused me on how this works.